### PR TITLE
[spirv] Translate several intrinsic functions.

### PR DESCRIPTION
**Pull request by ehsannas**
* frexp
* countbits
* clip
* ldexp
* lerp
* log10
* mad
* modf
* reversebits

### CODE REVIEW

**Comment by ehsannas** [comment]
PTAL @antiagainst

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp [hunk 1]
```diff
@@ -3422,6 +3437,259 @@ uint32_t SPIRVEmitter::processIntrinsicCallExpr(const CallExpr *callExpr) {
...
+}
+
+uint32_t SPIRVEmitter::processIntrinsicFrexp(const CallExpr *callExpr) {
+  // Signature is: ret frexp(x, exp)
+  // [in]   x: the input floating-point value.
+  // [out]  exp: the calculated exponent.
+  // [out]  ret: the calculated mantissa.
+  // All of the above must be a scalar, vector, or matrix of *float* type.
+
+  // The Frexp SPIR-V instruction returns a struct. The first
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: s/Frexp/FrexpStruct/

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp [hunk 2]
```diff
@@ -3422,6 +3437,259 @@ uint32_t SPIRVEmitter::processIntrinsicCallExpr(const CallExpr *callExpr) {
...
+  // Could not determine the argument as a constant. We need to branch based on
+  // the argument. If the argument is a vector/matrix, clipping is done if *any*
+  // element of the vector/matrix is less than zero.
+  const uint32_t argId = doExpr(arg);
+
+  QualType elemType = {};
+  uint32_t elemCount = 0, rowCount = 0, colCount = 0;
+  if (TypeTranslator::isScalarType(argType)) {
+    const auto zero = getValueZero(argType);
+    const auto cmpType = boolType;
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: Why not use `boolType` directly in the following? It seems there is no need to create a new  short lived `cmpType`.

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp [hunk 3]
```diff
@@ -3422,6 +3437,259 @@ uint32_t SPIRVEmitter::processIntrinsicCallExpr(const CallExpr *callExpr) {
...
+                                                 boolColType, lhsVec, zero);
+      const auto any = theBuilder.createUnaryOp(spv::Op::OpAny, boolType, cmp);
+      cmpResults.push_back(any);
+    }
+    const auto boolRowType = theBuilder.getVecType(boolType, rowCount);
+    const auto results =
+        theBuilder.createCompositeConstruct(boolRowType, cmpResults);
+    condition = theBuilder.createUnaryOp(spv::Op::OpAny, boolType, results);
+  } else {
+    emitError("Invalid type passed to clip function.");
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: return here?

Comment (ehsannas) [reply tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: Done

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp (hunk 2 above)
Comment (ehsannas) [reply tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: Done

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp (hunk 1 above)
Comment (ehsannas) [reply tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: Done

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp [hunk 4]
```diff
@@ -3422,6 +3437,259 @@ uint32_t SPIRVEmitter::processIntrinsicCallExpr(const CallExpr *callExpr) {
...
 
+uint32_t SPIRVEmitter::processIntrinsicModf(const CallExpr *callExpr) {
+  // Signature is: ret modf(x, ip)
+  // [in]    x: the input floating-point value.
+  // [out]  ip: the integer portion of x.
+  // [out] ret: the fractional portion of x.
+  // All of the above must be a scalar, vector, or matrix with the same
+  // component types. Component types can be float or int.
+
+  // The Frexp SPIR-V instruction returns a struct. The first member is the
```
[the remark this answers is not in the document]
Comment (ehsannas) [reply tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: Done

**Pull request merged**